File: 2130. Maximum Twin Sum of a Linked List.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func convertLinkedListToSlice(head *ListNode) []int {
	cur := head
	result := []int{}
	for cur != nil {
		result = append(result, cur.Val)
		cur = cur.Next
	}
	return result
}

func findTwinOfIndex(i, n int) int {
	return (n - 1) - i
}

func pairSum(head *ListNode) int {
	array := convertLinkedListToSlice(head)
	answer := -1
	for i := 0; i <= (len(array)/2 - 1); i++ {
		fmt.Println(i)
		answer = max(answer, array[i]+array[findTwinOfIndex(i, len(array))])
	}
	return answer
}

func main() {

	head := &ListNode{5, nil}
	l2 := &ListNode{4, nil}
	l3 := &ListNode{2, nil}
	l4 := &ListNode{1, nil}

	head.Next = l2
	l2.Next = l3
	l3.Next = l4

	answer := pairSum(head)

	fmt.Println("Answer: ", answer)

}

File: 2130. Maximum Twin Sum of a Linked List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:

        stack = collections.deque()
        result = -sys.maxsize
        cur = head

        # Create Stack
        while cur != None:
            stack.append(cur.val)
            cur = cur.next
        
        while head != None:
            result = max(result,head.val+stack.pop())
            head = head.next

        return result


    # Code is not the best, but it works. Its Linear Time & Constant Space
    # Came up with this solution on my own
    def LinearTimeConstantSpacepairSum(self, head: Optional[ListNode]) -> int:
        
        listSize = 0 
        cur = head

        while cur != None:
            cur = cur.next
            listSize += 1
        
        mid = int(listSize/2)
        cur = head
        n = 0

        while n != mid:
            cur = cur.next
            n += 1

        # Reverse List
        prev = None
        
        while cur != None:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp
        
        result = float('-inf')
        cur = head
        counter = 0

        while counter != mid:
            result = max(result,prev.val + cur.val)
            prev = prev.next
            cur = cur.next
            counter += 1


        return result